cfgrammar/src/lib/yacc/parser.rs

```diff
@@ -315,6 +327,15 @@ impl YaccParser {
             i = self.parse_rule(i)?;
             i = self.parse_ws(i, true)?;
         }
+
+        if let Some((_, spans)) = self.duplicate_rule_spans.iter().next() {
+            let span = spans.first().unwrap();
+            return Err(YaccParserError {
+                kind: YaccParserErrorKind::DuplicateRule,
+                span: *span,
+            });
+        }
```

Comment:
This checks whether there exists an error, and returns the first duplicate span,
as `YaccParserError` still just contains a single `kind`/`span`, as there isn't a way to return multiple errors yet.

Reply:
Yeah, the big question here is if we should do it on a per-`YaccParserErrorKind` basis, or on a `YaccParserError` basis.

Doing it in `YaccParserError` means, *any* error can produce multiple errors.
for instance, we can still perhaps produce syntax errors (e.g. `MissingRightArrow` after a `DuplicateRule` is encountered, and it may be nice to produce a  `vec![(DuplicateRule, span), (MissingRightArrow, span]`.

For this i'm kind of partial to changing `YaccParserError` to have something like a `Vec<YaccParserErrorKind, Span>`.
Making it part of `YaccParserError` also means, we don't have to change the semver compatibility for other `Duplicate*` errors like `DuplicatePrecedence`.

I think that both of these aspects (semver, and encountering multiple errors) make me lean towards keeping `DuplicateRule` as-is, and adding it to the `YaccParserError`.

Reply:
There are two different things here:

1. Grammars can have multiple errors.
2. A given rule can be duplicated more than once.

In this case, we're not yet handling #1 but we are handling #2. So `YaccParserErrorKind::DuplicateRule(Vec<Span>)` makes sense here IMHO.

Reply:
Right, the only qualm I have with this is that it isn't reusable for other `Duplicate*` types, of which there are quite a few.
So adding the `Vec<Span>` has to be done for each one of these individually as a semver breaking change.

Perhaps we could consider adding an `additional_spans: Vec<Span>` in `YaccParserError` to sit along side the `kind`?
The major difference being with this solution _every_ error, even ones which will never produce multiple spans, have an empty vector of additional spans.

Reply:
> So adding the Vec<Span> has to be done for each one of these individually as a semver breaking change.

Yes, that's true. Of course there's no reason why we can't add a `Vec` to them now, even if we don't really populate it. [Not in this PR! But in the next PR or two.]

> Perhaps we could consider adding an additional_spans: Vec<Span> in YaccParserError to sit along side the kind?

I'm not keen on this approach, because it only applies to a subset of errors.

Reply:
Yep!

Reply:
Fixed in https://github.com/softdevteam/grmtools/pull/307/commits/8cb0b6746e9b46b2b7ce860c63fa0337b24bbea2

I managed to salvage the macro, but it isn't exactly pretty.
The `spans.clone()` is kind of regrettable here, I think the only way to avoid that clone is to move it back out of `self`.